src/message/messagesReducer.js

```diff
@@ -17,15 +17,55 @@ import {
   EVENT_UPDATE_MESSAGE,
 } from '../actionConstants';
 import { NULL_ARRAY } from '../nullObjects';
+import { getNarrowsForMessage, keyFromNarrow } from '../utils/narrow';
 
 const initialState: MessagesState = Immutable.Map([]);
 
 const eventNewMessage = (state, action) => {
-  // TODO: Optimize -- Only update if the new message belongs to at least
-  // one narrow that is caught up.
+  const { message, caughtUp } = action;
+  const { flags } = message;
+
+  if (!flags) {
+    throw new Error('EVENT_NEW_MESSAGE message missing flags');
+  }
+
+  // Don't add a message that's already been added. It's probably
+  // very rare for a message to have already been added when we
+  // get an EVENT_NEW_MESSAGE, and perhaps impossible. (TODO:
+  // investigate?)
   if (state.get(action.message.id)) {
     return state;
   }
+
+  const narrowsForMessage = getNarrowsForMessage(message, action.ownUser, flags);
+  // `narrowsForMessage` will always have at least length 1 because of
+  // the home narrow.
+  const anyNarrowIsCaughtUp = narrowsForMessage.some(narrow => {
+    const key = keyFromNarrow(narrow);
+    // (No guarantee that `key` is in `action.caughtUp`)
+    // flowlint-next-line unnecessary-optional-chain:off
+    return caughtUp[key]?.newer;
+  });
+
+  // We only need to record the message if we intend to show it right
+  // away in any narrow, i.e., before the next MESSAGE_FETCH_COMPLETE.
+  // We only intend to show it right away in a narrow if we know it's
+  // that narrow's most recent message, i.e., we know we're currently
+  // looking at (caught up with) the newest messages in the narrow.
```

Comment:
I think it's not quite about whether we'll "show it right away" -- we could be caught up simply if the user has previously gone and looked at that narrow, and seen the latest messages in it, and so we've stayed caught up since then.

What's really going on is basically a converse to the warning added below about making sure this contains everything that's in `state.narrows`: it's that we don't really have a use for anything to be in `state.messages` if it *isn't* in `state.narrows`. The only times we try to access a message from here are when we've found its ID under some narrow in `state.narrows`.